File: backend/app/routers/mechanisms.py
"""SIM mechanism steps CRUD (PRD Epic 4)."""
from fastapi import APIRouter, Depends, status
from pymysql.connections import Connection

from ..database import get_db
from ..models import Mechanism

router = APIRouter(prefix="/api/mechanisms", tags=["mechanisms"])

_SELECT = (
    "SELECT id, requirement_name AS requirementName, issuing_body AS issuingBody, "
    "cost, notes FROM mechanisms ORDER BY sort_order, id"
)


@router.get("", response_model=list[Mechanism])
def list_mechanisms(db: Connection = Depends(get_db)):
    with db.cursor() as cur:
        cur.execute(_SELECT)
        return [Mechanism(**row) for row in cur.fetchall()]


@router.post("", response_model=Mechanism, status_code=status.HTTP_201_CREATED)
def create_mechanism(mechanism: Mechanism, db: Connection = Depends(get_db)):
    with db.cursor() as cur:
        cur.execute(
            "INSERT INTO mechanisms (requirement_name, issuing_body, cost, notes, sort_order) "
            "VALUES (%(requirementName)s, %(issuingBody)s, %(cost)s, %(notes)s, "
            "(SELECT COALESCE(MAX(sort_order), 0) + 1 FROM mechanisms m))",
            mechanism.model_dump(exclude={"id"}),
        )
        mechanism.id = cur.lastrowid
    return mechanism


@router.put("/{mechanism_id}", response_model=Mechanism)
def update_mechanism(mechanism_id: int, mechanism: Mechanism, db: Connection = Depends(get_db)):
    mechanism.id = mechanism_id
    with db.cursor() as cur:
        cur.execute(
            "UPDATE mechanisms SET requirement_name=%(requirementName)s, "
            "issuing_body=%(issuingBody)s, cost=%(cost)s, notes=%(notes)s WHERE id=%(id)s",
            mechanism.model_dump(),
        )
    return mechanism


@router.delete("/{mechanism_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_mechanism(mechanism_id: int, db: Connection = Depends(get_db)):
    with db.cursor() as cur:
        cur.execute("DELETE FROM mechanisms WHERE id=%(id)s", {"id": mechanism_id})

File: attendance-backend/handlers/mechanisms.go
package handlers

// SIM mechanism steps CRUD (PRD Epic 4). Migrated from the FastAPI service.
// Public (no auth), bare JSON bodies.

import (
	"net/http"

	"field-attendance-system/database"
	"field-attendance-system/models"

	"github.com/gin-gonic/gin"
)

func ListMechanisms(c *gin.Context) {
	mechs := make([]models.Mechanism, 0)
	if err := database.DB.Order("sort_order, id").Find(&mechs).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to list mechanisms"})
		return
	}
	c.JSON(http.StatusOK, mechs)
}

func CreateMechanism(c *gin.Context) {
	var mech models.Mechanism
	if err := c.ShouldBindJSON(&mech); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	// Auto-assign the next sort_order, mirroring the Python INSERT subquery.
	var maxOrder int
	database.DB.Model(&models.Mechanism{}).Select("COALESCE(MAX(sort_order), 0)").Scan(&maxOrder)
	mech.ID = 0
	mech.SortOrder = maxOrder + 1
	if err := database.DB.Create(&mech).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create mechanism"})
		return
	}
	c.JSON(http.StatusCreated, mech)
}

func UpdateMechanism(c *gin.Context) {
	id, err := parseUintParam(c, "id")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid id"})
		return
	}
	var mech models.Mechanism
	if err := c.ShouldBindJSON(&mech); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	mech.ID = id
	// Update only the four wire fields — never sort_order (a Save would zero it
	// and scramble list order). A map updates only the listed columns.
	if err := database.DB.Model(&models.Mechanism{}).Where("id = ?", id).Updates(map[string]interface{}{
		"requirement_name": mech.RequirementName,
		"issuing_body":     mech.IssuingBody,
		"cost":             mech.Cost,
		"notes":            mech.Notes,
	}).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update mechanism"})
		return
	}
	c.JSON(http.StatusOK, mech)
}

func DeleteMechanism(c *gin.Context) {
	id, err := parseUintParam(c, "id")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid id"})
		return
	}
	if err := database.DB.Delete(&models.Mechanism{}, id).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete mechanism"})
		return
	}
	c.Status(http.StatusNoContent)
}
